File: lib-segment/src/main/java/websiteschema/mpsegment/core/SegmentEngine.java
package websiteschema.mpsegment.core;

import websiteschema.mpsegment.conf.MPSegmentConfiguration;
import websiteschema.mpsegment.dict.DictionaryFactory;

import java.util.HashMap;
import java.util.Map;

public class SegmentEngine {

    private final static ThreadLocal<SegmentWorker> workers = new ThreadLocal();
    private final static SegmentEngine instance = new SegmentEngine();
    private final MPSegmentConfiguration configuration = MPSegmentConfiguration.getInstance();

    private SegmentEngine() {
        loadDictionary();
    }

    private void loadDictionary() {
        DictionaryFactory.getInstance().loadDictionary();
        DictionaryFactory.getInstance().loadDomainDictionary();
        DictionaryFactory.getInstance().loadUserDictionary();
    }

    public static SegmentEngine getInstance() {
        return instance;
    }

    public SegmentWorker getReusableSegmentWorker() {
        SegmentWorker segmentWorker = workers.get();
        if (null == segmentWorker) {
            segmentWorker = new SegmentWorker(configuration);
            workers.set(segmentWorker);
        }

        return segmentWorker;
    }

    public SegmentWorker getSegmentWorker() {
        return new SegmentWorker(configuration);
    }

    public SegmentWorker getSegmentWorker(Map<String,String> config) {
        return new SegmentWorker(new MPSegmentConfiguration(config));
    }

    public SegmentWorker getSegmentWorker(String ... props) {
        Map<String, String> map = null;
        if(null != props) {
            map = new HashMap<String, String>(props.length);
            for(String p : props) {
                String[] keyAndValue = p.split("(=|->)");
                if(null != keyAndValue && keyAndValue.length == 2) {
                    map.put(keyAndValue[0].trim(), keyAndValue[1].trim());
                }
            }
        }
        return getSegmentWorker(map);
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/core/SegmentEngine.scala
package websiteschema.mpsegment.core

import websiteschema.mpsegment.conf.MPSegmentConfiguration
import websiteschema.mpsegment.dict.DictionaryFactory

import collection.mutable.{Map, HashMap}

class SegmentEngine {

  private val configuration = MPSegmentConfiguration()

  implicit def javaMapToScalaMap(javaMap: java.util.Map[String, String]) = {
    val map = HashMap[String, String]()
    val iterator = javaMap.entrySet().iterator()
    while (iterator.hasNext) {
      val entry = iterator.next()
      map += (entry.getKey -> entry.getValue)
    }
    map
  }

  private def loadDictionary() {
    DictionaryFactory().loadDictionary()
    DictionaryFactory().loadDomainDictionary()
    DictionaryFactory().loadUserDictionary()
    DictionaryFactory().loadEnglishDictionary()
  }

  def getSegmentWorker(): SegmentWorker = {
    return new SegmentWorker(configuration)
  }

  def getSegmentWorker(config: java.util.Map[String, String]): SegmentWorker = {
    return new SegmentWorker(MPSegmentConfiguration(config))
  }

  def getSegmentWorker(props: String*): SegmentWorker = {
    var map: Map[String, String] = null
    if (null != props) {
      map = HashMap[String, String]()
      for (p <- props)
      {
        val keyAndValue = p.split("(=|->)")
        if (null != keyAndValue && keyAndValue.length == 2) {
          map += (keyAndValue(0).trim() -> keyAndValue(1).trim())
        }
      }
    }
    return new SegmentWorker(MPSegmentConfiguration(map))
  }
}

object SegmentEngine {
  val instance = new SegmentEngine()
  instance.loadDictionary()
  def apply() = instance
}
